### PR TITLE
ellipsis for typesafe variadic parameters not shown

### PR DESCRIPTION
**Pull request by MartinNowak**
```d
module app;

///
void bug(int[4] args...)
{
}

///
void bug(int[] args...)
{
}
```
Shown as
```
void bug(
  int[4] args
);

void bug(
  int[] args
); 
```
What's a bit tricky here is that this "styling" of the parameter depends on the deco of the function. Hopefully not too difficult to transport the information.

### CODE REVIEW

File: views/ddox.inc.function.dt [hunk 1]
```diff
@@ -31,7 +32,7 @@
 			br
 			- foreach (size_t i, p; proto.parameters)
 				- auto pattribs = highlightDCode(getAttributeString(p.attributes, AttributeStringKind.normal));
-				- auto suffix = i+1 < proto.parameters.length ? "<span class=\"pun\">,</span>" : "";
+				- auto suffix = i+1 < proto.parameters.length ? "<span class=\"pun\">,</span>" : variadic_suffix;
```
Comment (MartinNowak) [inline views/ddox.inc.function.dt]: I just attached it to the last parameter, even though `, …` could arguably be put after a `br`, but that makes the formatting a lot more complex.

File: source/ddox/api.d [hunk 2]
```diff
@@ -350,6 +352,16 @@ void formatType(R)(ref R dst, CachedType type, scope string delegate(in Entity)
 	if (include_code_tags) dst.put("</code>");
 }
 
+string getVariadicSuffix(Type type)
+{
+	if (type.typesafeVariadic)
+		return "…";
+	else if (type.variadic)
+		return type.parameterTypes.length ? ", …" : "…";
```
Comment (MartinNowak) [inline source/ddox/api.d]: Not sure about using ellipsis or dots.

File: source/ddox/entities.d [hunk 3]
```diff
@@ -612,6 +612,7 @@ struct Type {
 	immutable(CachedType)[] parameterTypes;
 	immutable(CachedString)[] _parameterNames;
 	immutable(Value)[] _parameterDefaultValues;
+	bool variadic, typesafeVariadic;
```
Comment (s-ludwig) [inline source/ddox/entities.d]: Would be a bt nicer to use an `enum VariadicKind { none, generic, typeSafe }` here.

Comment (MartinNowak) [reply source/ddox/entities.d]: Done, just used [std.traits.Variadic](https://dlang.org/library/std/traits/variadic.html).

**Pull request merged**